fix local name customization for usercss/legacy

tophf opened the pull request:
Should fix #1045.

The UI is simple: the same old name input and an `x` button next to the name shown when a custom name is entered that differs from the original name, tooltip: "Reset custom name to the original style name"

* UserCSS editor: the name input now accepts editing
  * the value always goes into `customName`
  * when emptied manually it shows the default placeholder ("Specify @name in the code") which hopefully won't be too confusing since it'll be automatically reset to the real `@name` on saving or editing the source text
* Legacy editor: the name input value goes either in `customName` (for updatable styles) or in `name` (for locally created styles)
* Popup and manager display `customName` if it's present
* ~If `customName` is equal to `name`, it's automatically removed on saving in the editor~

eight04 reviewed (dismissed):
1. Do we search `customName` property in manager filter?
2. For the UI, I think it is better to use two different inputs. One for `name` and one for `customName`. It also helps us managing the dirty state.
3. We should show why a user should use custom name. For example: https://codepen.io/eight04/pen/wvWwvaZ
4. I got an error that prevents the manager from working: 
![image](https://user-images.githubusercontent.com/1324510/95062499-a2231e80-072f-11eb-97d8-bc0cd1f2ba18.png)

tophf commented on background/style-manager.js:
```diff
@@ -380,6 +380,9 @@ const styleManager = (() => {
     if (!style.name) {
       throw new Error('style name is empty');
     }
+    if (!style.customName || style.customName === style.name) {
```
Yes, I knew about this case but I don't think it'll be a problem even if it happens once a decade to one user out of a million.

eight04 replied:
Then let's move this logic to the editor. To "reset" a custom name, we can assign `null` to it so it will overwrite the old value.

tophf replied:
Done. Hopefully I understood you correctly.

tophf commented:
> 1. Do we search `customName` property in manager filter?

Done.

> 2. For the UI, I think it is better to use two different inputs. One for `name` and one for `customName`. It also helps us managing the dirty state.

No, it's an unnecessary complication and we've discussed this in the linked issue already. There's no problem in managing the dirty state either.

> 3. We should show why a user should use custom name. For example: https://codepen.io/eight04/pen/wvWwvaZ

Done: added a tooltip.

> 4. I got an error that prevents the manager from working:

Fixed in master: e7a6e86b

tophf commented on edit/edit.js:
```diff
@@ -193,16 +194,30 @@ preinit();
+          .then(updateDirty)
+          .then(!usercss && (() => {
+            // updating after init to prevent flash of validation failure on an empty name
+            $('#name').required = true;
```
No, I think it'd be better to fix the editor to restore its initial speed that was lost during one of the storage refactors.

eight04 replied:
So we revert this change and improve initial speed later?

tophf replied:
No, we move on for now because no one knows how long it'll take.

tophf replied:
It's probably something that you or someone familiar with tooling should kick off because one of the biggest contributors to the slow loading of our editor per Chrome's devtools profiler is that it loads a lot of small files and that alone wastes 100ms just for the network transfer. This is a ridiculously long time. A single bundle or two in a typical production page would load in a few milliseconds.

This problem is not evident in Firefox because it suppresses first paint when the page is opened. On the one hand it makes sense for many poorly built pages like ours, but on the other hand it feels much slower than Chrome which manages to show several intermediate frames [useless though] while loading.

Bundling+minifying would be also useful for our CSSLint because editing two huge files is very inconvenient. Maybe we could start with that e.g. move it to a project in `openstyles/csslint-se` where `se` is "stylus edition".

tophf replied:
As for why I'm against the `body` class, toggling it invalidates the layout for the entire DOM tree which requires expensive calculation. It was one of the reasons I wrote a custom keyboard-accessibility-focus implementation instead of the popular one that modified an attribute on `<html>` or `<body>`.

eight04 commented:
> No, it's an unnecessary complication and we've discussed this in the linked issue already. There's no problem in managing the dirty state either.

Here is an example: 
![test](https://user-images.githubusercontent.com/1324510/95116537-c277cb00-0779-11eb-8eca-914659d6c5a7.gif)

tophf commented:
>Here is an example:

I guess I'm not a visual person. I don't understand what happens there.

dnknn commented:
@tophf I have a good idea, we can do the same as `Tampermonkey` , WDYT?

![2010_06075554](https://user-images.githubusercontent.com/10995302/95143716-6331af00-07a9-11eb-80f7-2d68f6455348.png)



`// @name`　　 Original name, don't change it, because it is an API interface.
　　 　　 But you can use the following localized language names to customize the display.

`// @name:en`
`// @name:jp`
`// @name:zh`	
`// @name:zh_tw`
.....
.....

tophf commented:
@dnknn it's a good idea but entirely different, already reported in #377.

tophf commented:
@eight04 so I've watched your gif like 20 times over the last 5 days and I still can't understand what happens there. If you found a bug could you please write the repro sequence as text? Otherwise I'm merging this PR and let the users find the bugs. It's not like this feature is used frequently anyway.

dnknn commented:
I help tophf ≺at≻ @eight04

tophf commented:
Ah, I get it now. I need to store the states differently.

tophf commented:
I've reworked it so now both the code and the user's workflow should be more self-explanatory. The entered custom name will always be kept even if it matches the current style's inherent name. The `x` button to reset it will be always shown next to it. Clicking the button will remove the custom name from the style (and database), the `x` button will hide.

eight04 commented:
> merging this PR

I'm not able to work on code these days (and probably next week) so you may want to merge them first if needed.

tophf commented:
Ah, I see. Well, I'll test the rewrite for a few days myself first anyway.

narcolepticinsomniac commented:
There must be some reason `#message-box-contents` has `tabindex="0"` set, but I haven't got a clue what purpose it serves.

narcolepticinsomniac commented:
Editors are impressively quick. Only issue I still see occasionally, albeit less frequently than before, is remove section buttons being slow. When I said I'm testing monster multi-sectioned styles, these are a couple where the first section is the majority of like 200kb styles. I assume that's the reason, which is understandable.

tophf commented:
> remove section buttons being slow

I didn't encounter that yet so next time it happens maybe you could record a trace in devtools `Performance` panel. Open it, press Ctrl-E to start recording, do something to reproduce the problem, switch back to the panel and press Ctrl-E to stop recording. It'll build a flamechart where the culprit may be immediately obvious, otherwise save the trace (there's a button in the toolbar), and upload a zip of it here. FWIW Instead of Ctrl-E you can press the first button in the toolbar.

eight04 commented on edit/edit.js:
```diff
@@ -88,27 +120,10 @@ preinit();
-      });
-    }
+    CODEMIRROR_THEMES.unshift(chrome.i18n.getMessage('defaultTheme'));
+    $('#editor.theme').append(...CODEMIRROR_THEMES.map(s => $create('option', s)));
```
We may want to mention this change in changelog or something that we no longer read codemirror themes from local files.

tophf replied:
Only if you do it yourself :-) I would wait for a report from a real user who was inventive enough to put custom files there.
